drivers/spi/spi_bitbang.c

```diff
@@ -0,0 +1,301 @@
+	struct gpio_dt_spec mosi_gpio;
+	struct gpio_dt_spec miso_gpio;
+};
+
+static inline bool spi_bitbang_has_miso(const struct device *dev)
+{
+	const struct spi_bitbang_config *info = dev->config;
+
+	return info->miso_gpio.port != NULL;
+}
+
+static inline bool spi_bitbang_has_mosi(const struct device *dev)
+{
+	const struct spi_bitbang_config *info = dev->config;
+
+	return info->mosi_gpio.port != NULL;
+}
+
+static int spi_bitbang_configure(const struct spi_bitbang_config *info,
+			    struct spi_bitbang_data *data,
+			    const struct spi_config *config)
+{
+	if (config->operation & SPI_OP_MODE_SLAVE) {
+		LOG_ERR("Slave mode not supported");
+		return -ENOTSUP;
+	}
+
+	if (config->operation & (SPI_TRANSFER_LSB | SPI_LINES_DUAL
+			| SPI_LINES_QUAD)) {
+		LOG_ERR("Unsupported configuration");
+		return -ENOTSUP;
+	}
+
+	const int bits = SPI_WORD_SIZE_GET(config->operation);
+
+	if (bits > 16) {
+		LOG_ERR("Word sizes > 16 bits not supported");
+		return -ENOTSUP;
+	}
+
+	data->bits = bits;
+	data->dfs = ((data->bits - 1) / 8) + 1;
+	if (config->frequency > 0) {
+		/* convert freq to period, the extra /2 is due to waiting
+		 * twice in each clock cycle. The '2000' is an upscale factor.
+		 */
+		data->wait_us = (1000000ul * 2000ul / config->frequency) / 2000ul;
+		data->wait_us /= 2;
+	} else {
+		data->wait_us = 8 / 2; /* 125 kHz */
+	}
+
+	data->ctx.config = config;
+
+	spi_context_cs_configure(&data->ctx);
+
+	return 0;
+}
+
+static int spi_bitbang_transceive(const struct device *dev,
+			      const struct spi_config *spi_cfg,
+			      const struct spi_buf_set *tx_bufs,
+			      const struct spi_buf_set *rx_bufs)
+{
+	const struct spi_bitbang_config *info = dev->config;
+	struct spi_bitbang_data *data = dev->data;
+	struct spi_context *ctx = &data->ctx;
+
+	spi_bitbang_configure(info, data, spi_cfg);
+
+	spi_context_buffers_setup(ctx, tx_bufs, rx_bufs, data->dfs);
+
+	int clock_state = 0;
+	int cpha = 0;
+	bool loop = false;
+
+	if (SPI_MODE_GET(spi_cfg->operation) & SPI_MODE_CPOL) {
+		clock_state = 1;
+	}
+	if (SPI_MODE_GET(spi_cfg->operation) & SPI_MODE_CPHA) {
+		cpha = 1;
+	}
+	if (SPI_MODE_GET(spi_cfg->operation) & SPI_MODE_LOOP) {
+		loop = true;
+	}
+
+	/* set the initial clock state before CS */
+	gpio_pin_set_dt(&info->clk_gpio, clock_state);
+
+	spi_context_cs_control(ctx, true);
+
+	const uint32_t wait_us = data->wait_us;
+
+	const bool has_miso = spi_bitbang_has_miso(dev);
+	const bool has_mosi = spi_bitbang_has_mosi(dev);
+
+	while (spi_context_tx_buf_on(ctx) || spi_context_rx_buf_on(ctx)) {
+		uint16_t w = 0;
+
+		if (ctx->tx_len) {
+			switch (data->dfs) {
+			case 2:
+				w = *(uint16_t *)(ctx->tx_buf);
+				break;
+			case 1:
+				w = *(uint8_t *)(ctx->tx_buf);
+				break;
+			}
+		}
+
+		int shift = data->bits - 1;
+		uint16_t r = 0;
+		int b = 0;
+		bool do_read = false;
+
+		if (has_miso && spi_context_rx_buf_on(ctx)) {
+			do_read = true;
+		}
+
+		while (shift >= 0) {
+			const int d = (w >> shift) & 0x1;
```

Comment:
Since this is only an MSB first transfer... shouldn't it be checking that it's not being configured for SPI_TRANSFER_LSB.